calico/felix/devices.py

```diff
@@ -315,46 +322,68 @@ def watch_interfaces(self):
-                    if msg_type == RTM_NEWLINK:
-                        _log.debug("Detected new network interface : %s",
-                                   rta_data)
-                        self.update_splitter.on_interface_update(rta_data,
-                                                                 async=True)
-                    else:
-                        _log.debug("Network interface has gone away : %s",
-                                   rta_data)
+            _log.debug("Netlink message type %s len %s", msg_type, msg_len)
+
+            if msg_type in [RTM_NEWLINK, RTM_DELLINK]:
+                # A new or removed interface.  Read the struct
+                # ifinfomsg, which is 16 bytes.
+                hdr = data[:16]
+                data = data[16:]
+                _, _, _, index, flags, _ = struct.unpack("=BBHiII", hdr)
+                _log.debug("Interface index %s flags %x", index, flags)
+
+                # Bytes left is the message length minus the two headers of 16
+                # bytes each.
+                remaining = msg_len - 32
+
+                # Loop through attributes, looking for the pieces of
+                # information that we need.
+                ifname = None
+                operstate = None
+                while remaining:
+                    # The data content is an array of RTA objects, each of
+                    # which has a 4 byte header and some data.
+                    rta_len, rta_type = struct.unpack("=HH", data[:4])
+
+                    # This check comes from RTA_OK, and terminates a string of
+                    # routing attributes.
+                    if rta_len < 4:
+                        break
+
+                    rta_data = data[4:rta_len]
+
+                    # Remove the RTA object from the data. The length to jump
+                    # is the rta_len rounded up to the nearest 4 byte boundary.
+                    increment = int((rta_len + 3) / 4) * 4
+                    data = data[increment:]
+                    remaining -= increment
+
+                    if rta_type == IFLA_IFNAME:
+                        ifname = rta_data[:-1]
+                        _log.debug("IFLA_IFNAME: %s", ifname)
+                    elif rta_type == IFLA_OPERSTATE:
+                        operstate, = struct.unpack("=B", rta_data[:1])
+                        _log.debug("IFLA_OPERSTATE: %s", operstate)
+
+                if (ifname and
+                    msg_type == RTM_DELLINK and
+                    ifname in if_last_flags):
+                    # An interface that we've previously signalled is
+                    # being deleted, so remove our record of it.
+                    del if_last_flags[ifname]
+
+                if (ifname and
+                    msg_type == RTM_NEWLINK and
+                    operstate == IF_OPER_UP and
+                    (ifname not in if_last_flags or
+                     if_last_flags[ifname] != flags)):
```

Comment:
Hmm, I suppose we don't store it until we see it oper up, so maybe the latter is OK.  Can you confirm that going oper down isn't enough to remove the routes and require reprogramming?